### PR TITLE
Scale down on allocation exception only and other fixes

### PR DESCRIPTION
**Pull request by alena1108**

### CODE REVIEW

**Comment by alena1108** [comment]
https://github.com/rancher/rancher/issues/5239

File: code/iaas/service-discovery/server/src/main/java/io/cattle/platform/servicediscovery/deployment/impl/unit/DefaultDeploymentUnitInstance.java [hunk 1]
```diff
@@ -229,14 +232,18 @@ public boolean isHealthCheckInitializing() {
...
+                instance = context.resourceMonitor.waitFor(instance, new ResourcePredicate<Instance>() {
+                    @Override
+                    public boolean evaluate(Instance obj) {
+                        return context.objectManager.find(InstanceHostMap.class, INSTANCE_HOST_MAP.INSTANCE_ID,
+                                instance.getId()).size() > 0;
+                    }
+                });
+            }
+        } catch (Exception ex) {
+            throw new ServiceInstanceAllocateException("Failed to allocate instance [" + instance.getId() + "]");
```
Comment (ibuildthecloud) [inline code/iaas/service-discovery/server/src/main/java/io/cattle/platform/servicediscovery/deployment/impl/unit/DefaultDeploymentUnitInstance.java]: Probably don't want to mask the original exception.  Can you pass it to the constructor?

Comment (alena1108) [inline code/iaas/service-discovery/server/src/main/java/io/cattle/platform/servicediscovery/deployment/impl/unit/DefaultDeploymentUnitInstance.java]: will do

**Comment by alena1108** [comment]
Also added the fix to get rid of DataBaseChangedExceptions to this PR

**Pull request merged**